### PR TITLE
feat: add Lance filter API

### LINKED ISSUE #21: feat: support filtering datasets by sample id
## Problem

Some pretraining datasets need to include or exclude samples by a predefined list of sample ids.

For example, an experiment may use:

- `include_ids_file`: only keep samples whose `id` is listed in the file
- `exclude_ids_file`: drop samples whose `id` is listed in the file

This is useful for train/validation split control, deduplication, and reproducing existing experiment configs.

Currently, `mvp-dataset` has related dataset operations such as:

- `select(fields)`: field projection, not sample filtering
- `sample(fraction)`: random subset by fraction
- `split(fractions)`: row-based split
- Lance `_selection`: row-window / sampled-row selection
- `assemble(...)`: can be used to implement a lazy post-read filter manually

But there is no first-class operation for filtering samples by id.

## Requested Feature

Add a dataset filtering op that supports id-based include/exclude semantics.

Suggested API shape:

```python
dataset = Dataset.from_source(...)
dataset = dataset.filter_by_id(
    id_field="id",
    include_ids_file="/path/to/include.txt",
    exclude_ids_file="/path/to/exclude.txt",
)
```

or a more general filter API:

```python
dataset = Dataset.from_source(...)
dataset = dataset.filter(predicate)
```

with a built-in helper for id files.

## Desired Semantics

- `include_ids_file`: keep only samples whose `id_field` is in the file
- `exclude_ids_file`: drop samples whose `id_field` is in the file
- If both are provided, apply include first, then exclude
- The id file should contain one id per line
- Empty lines should be ignored
- The filter config should be included in dataset fingerprint / resume compatibility checks
- Missing `id_field` should raise a clear error

## Lance-Specific Consideration

For Lance datasets, a source-level implementation would be preferable when possible:

1. Resolve ids to row positions
2. Cache the id-to-row-index mapping or selected row positions
3. Read only selected rows during iteration

This avoids the inefficiency of post-read filtering, especially when the include set is small or the exclude set is large.

A simpler first implementation could be a lazy post-read filter stage, but the long-term goal should be source-level filtering for Lance.

## Motivation

This is needed to reproduce LLM pretraining configs that rely on train/validation split files, such as excluding validation ids from the training dataset and evaluating only on a fixed validation id set.

### PR DESCRIPTION
**Pull request by GeoffreyChen777**
## Summary

- add a lazy Lance-native `filter()` API for SQL predicates and batched predicate lists
- build a disk-backed filtered-row index cooperatively across ranks and DataLoader workers
- preserve source order, deduplicate overlapping predicate batches, and support existing shuffle, split, sample, resume, and reference resolution paths
- document bounded `IN (...)` batches for ID-file filtering

## Why

Large ID filters should use Lance scalar indexes without constructing one oversized SQL predicate or materializing the filtered row mapping in memory. Predicate lists are OR-combined within one call, while repeated `filter()` calls remain AND-combined.

## Validation

- `pre-commit`: ruff check, ruff format, isort
- `pytest tests/test_lance_filter.py -q`: 25 passed
- full available suite: 146 passed, 16 skipped, 2 Torch-dependent tests deselected

## Configuration impact

The filter index defaults to shared scope. Remote or read-only Lance sources require `MVP_DATASET_LANCE_FILTER_INDEX_CACHE_DIR`.

close #21

### CODE REVIEW

**Comment by GeoffreyChen777** [comment]
## Example: include ID files with the filter API

The new predicate-list support can be used to implement ID-file inclusion without adding a dedicated `filter_by_id_files()` API. Read IDs from one or more files, deduplicate them, split them into bounded `IN (...)` predicates, and pass the complete predicate list to `filter()`.

```python
from collections.abc import Sequence
from pathlib import Path

from mvp_dataset import Dataset


def include_id_files(
    dataset: Dataset,
    id_files: Sequence[str | Path],
    *,
    id_column: str = "id",
    batch_size: int = 512,
) -> Dataset:
    ids: list[str] = []
    seen: set[str] = set()

    for path in id_files:
        with Path(path).open(encoding="utf-8") as file:
            for line in file:
                value = line.strip()
                if value and value not in seen:
                    seen.add(value)
                    ids.append(value)

    if not ids:
        return dataset.filter("FALSE")

    predicates = []
    for start in range(0, len(ids), batch_size):
        values = ", ".join(
            "'" + value.replace("'", "''") + "'"
            for value in ids[start : start + batch_size]
        )
        predicates.append(f"{id_column} IN ({values})")

    return dataset.filter(
        predicates,
        index={"scope": "shared"},
    )
```

Usage:

```python
dataset = Dataset.from_source(
    "lance",
    "/data/train.lance",
    columns=["id", "messages"],
)

dataset = include_id_files(
    dataset,
    [
        "/data/include-part-000.txt",
        "/data/include-part-001.txt",
    ],
)
```

Entries in the predicate list are logically OR-combined, but each entry is executed as an independent bounded Lance query instead of constructing one oversized SQL expression. Overlapping matches are deduplicated and results retain Lance source order.

This example assumes a string ID column and escapes SQL string literals. Integer ID columns should emit unquoted integer literals. A Lance scalar/BTree index on the ID column is strongly recommended; otherwise each predicate may require a scan. Missing IDs simply produce no rows, and cache preparation remains lazy until iteration or another operation needs the filtered row count.

**Review by flymin — APPROVED** [review]
LGTM! 

One non-blocking follow-up: it would be great to support a unified environment variable, such as `MVP_DATASET_CACHE_ROOT`, for configuring all cache paths at once.

**Comment by GeoffreyChen777** [comment]
> LGTM!
> 
> One non-blocking follow-up: it would be great to support a unified environment variable, such as `MVP_DATASET_CACHE_ROOT`, for configuring all cache paths at once.

That is exactly what I'm doing now:

image
